File: generated_code/GeminiPro/Python/007_SplayTree.py
class SplayTree:
    """
    A complete, self-contained implementation of a Splay Tree.

    This class acts as a dictionary-like set for storing unique integer keys.
    Splay trees are self-balancing binary search trees with the property that
    recently accessed elements are quick to access again. This is achieved by
    moving any accessed, inserted, or deleted node to the root of the tree
    through a series of rotations, an operation called "splaying".
    """

    class _Node:
        """A private inner class representing a node in the Splay Tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y

        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y

        y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splay operation on node x, moving it to the root.
        This operation consists of a series of rotations (Zig, Zig-Zig, Zig-Zag)
        until the node x is the root of the tree.
        """
        while x.parent:
            parent = x.parent
            grandparent = parent.parent
            if not grandparent:
                # Zig case
                if x == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif x == parent.left and parent == grandparent.left:
                # Zig-Zig case (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif x == parent.right and parent == grandparent.right:
                # Zig-Zig case (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif x == parent.right and parent == grandparent.left:
                # Zig-Zag case (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:
                # Zig-Zag case (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree and splays the accessed node.

        If the key is found, the node containing the key is splayed to the root.
        If the key is not found, the last non-null node visited during the
        search (the would-be parent) is splayed to the root.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        current = self.root
        last_node = None
        while current:
            last_node = current
            if key == current.key:
                self._splay(current)
                return True
            elif key < current.key:
                current = current.left
            else:
                current = current.right

        # Key not found, splay the last visited node if the tree was not empty
        if last_node:
            self._splay(last_node)

        return False

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the node with that key is splayed to the root.
        If the key is new, it is inserted as a new node and then that new node
        is splayed to the root.

        Args:
            key: The integer key to insert.
        """
        # Case 1: Tree is empty
        if not self.root:
            self.root = self._Node(key)
            return

        # Splay the closest node to the key's insertion point
        self.search(key)

        # Case 2: Key already exists, search has splayed it.
        if self.root.key == key:
            return

        # Case 3: Key does not exist, insert it and make it the new root.
        # The old root becomes a child of the new node.
        new_node = self._Node(key)
        if key < self.root.key:
            new_node.right = self.root
            new_node.left = self.root.left
            if self.root.left:
                self.root.left.parent = new_node
            self.root.left = None
        else:  # key > self.root.key
            new_node.left = self.root
            new_node.right = self.root.right
            if self.root.right:
                self.root.right.parent = new_node
            self.root.right = None

        self.root.parent = new_node
        self.root = new_node

    def delete(self, key):
        """
        Deletes a key from the tree.

        The node with the given key (or its would-be parent) is first splayed
        to the root. If the key is present, the node is removed, and the
        remaining subtrees are joined.

        Args:
            key: The integer key to delete.
        """
        if not self.root:
            return

        # Splay the node to be deleted (or its parent) to the root
        self.search(key)

        # If the key is not in the tree, the root after splaying will not have the key
        if self.root.key != key:
            return

        # The node to be deleted is now at the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # Promote the right subtree to be the new tree
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        else:
            # Disconnect the left subtree from the old root
            left_subtree.parent = None

            # Find the maximum node in the left subtree
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right

            # Splay this maximum node to the root of the left subtree.
            # We temporarily set self.root to the left subtree's root to reuse _splay.
            self.root = left_subtree
            self._splay(max_node)

            # After splaying, max_node is the new root. Its right child is None.
            # We attach the original right subtree to it.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root

File: generated_code/GeminiPro/Java/007_SplayTree.java
/**
 * A complete, self-contained Java class that implements an integer set using a Splay Tree.
 * A splay tree is a self-balancing binary search tree with the additional property that
 * recently accessed elements are quick to access again. All major operations (insert,
 * delete, search) involve a "splaying" step, which moves the accessed node to the root.
 */
public class SplayTree {

    /**
     * Inner class representing a node in the splay tree.
     */
    private class Node {
        int key;
        Node parent;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.parent = null;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty SplayTree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a left rotation on the given node x.
     *
     * <pre>
     *   x           y
     *  / \         / \
     * A   y  -->  x   C
     *    / \     / \
     *   B   C   A   B
     * </pre>
     *
     * @param x The node to rotate.
     */
    private void leftRotate(Node x) {
        Node y = x.right;
        if (y != null) {
            x.right = y.left;
            if (y.left != null) {
                y.left.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.left) {
                x.parent.left = y;
            } else {
                x.parent.right = y;
            }
            y.left = x;
            x.parent = y;
        }
    }

    /**
     * Performs a right rotation on the given node x.
     *
     * <pre>
     *     y         x
     *    / \       / \
     *   x   C --> A   y
     *  / \           / \
     * A   B         B   C
     * </pre>
     *
     * @param x The node to rotate.
     */
    private void rightRotate(Node x) {
        Node y = x.left;
        if (y != null) {
            x.left = y.right;
            if (y.right != null) {
                y.right.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.right) {
                x.parent.right = y;
            } else {
                x.parent.left = y;
            }
            y.right = x;
            x.parent = y;
        }
    }

    /**
     * Moves the given node to the root of the tree using a series of rotations.
     * This is the core operation of a splay tree.
     *
     * @param x The node to splay.
     */
    private void splay(Node x) {
        while (x.parent != null) {
            Node parent = x.parent;
            Node grandParent = parent.parent;
            if (grandParent == null) { // Zig step
                if (x == parent.left) {
                    rightRotate(parent);
                } else {
                    leftRotate(parent);
                }
            } else { // Zig-Zig or Zig-Zag step
                if (x == parent.left) {
                    if (parent == grandParent.left) { // Zig-Zig (left-left)
                        rightRotate(grandParent);
                        rightRotate(parent);
                    } else { // Zig-Zag (right-left)
                        rightRotate(parent);
                        leftRotate(grandParent);
                    }
                } else {
                    if (parent == grandParent.left) { // Zig-Zag (left-right)
                        leftRotate(parent);
                        rightRotate(grandParent);
                    } else { // Zig-Zig (right-right)
                        leftRotate(grandParent);
                        leftRotate(parent);
                    }
                }
            }
        }
    }

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the node with that key is splayed to the root.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        Node current = this.root;
        Node parent = null;
        while (current != null) {
            parent = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key already exists, splay the existing node and return.
                splay(current);
                return;
            }
        }

        Node newNode = new Node(key);
        newNode.parent = parent;

        if (parent == null) {
            this.root = newNode; // Tree was empty
        } else if (key < parent.key) {
            parent.left = newNode;
        } else {
            parent.right = newNode;
        }

        // Splay the newly inserted node to the root.
        splay(newNode);
    }

    /**
     * Searches for a key in the splay tree.
     * If the key is found, the corresponding node is splayed to the root and the key is returned.
     * If the key is not found, the last accessed node in the search path is splayed to the root
     * and null is returned.
     *
     * @param key The key to search for.
     * @return The key as an Integer if found, otherwise null.
     */
    public Integer search(int key) {
        Node nodeToSplay = findNode(key);
        if (nodeToSplay != null) {
            splay(nodeToSplay);
            // After splaying, the node is at the root. Check if its key matches.
            if (this.root.key == key) {
                return this.root.key;
            }
        }
        return null;
    }
    
    /**
     * Helper method to find a node or the last visited node on the search path.
     * This is used to determine which node to splay after an access.
     *
     * @param key The key to find.
     * @return The node containing the key, or the last visited node if not found.
     *         Returns null if the tree is empty.
     */
    private Node findNode(int key) {
        Node current = this.root;
        Node lastNode = null;
        while (current != null) {
            lastNode = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                return current; // Found the key
            }
        }
        return lastNode; // Return last visited node if key not found
    }

    /**
     * Deletes a key from the splay tree.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        // First, splay the node (or its closest predecessor/successor) to the root.
        Node nodeToSplay = findNode(key);
        if (nodeToSplay == null) {
            return; // Tree is empty
        }
        splay(nodeToSplay);

        // If the key is not at the root after splaying, it's not in the tree.
        if (this.root.key != key) {
            return;
        }

        // The node to delete is now the root.
        Node leftSubtree = this.root.left;
        Node rightSubtree = this.root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree.
            this.root = rightSubtree;
            if (this.root != null) {
                this.root.parent = null;
            }
        } else {
            // Disconnect the left subtree to treat it as its own tree.
            leftSubtree.parent = null;

            // Find the maximum node in the left subtree.
            Node maxInLeft = leftSubtree;
            while (maxInLeft.right != null) {
                maxInLeft = maxInLeft.right;
            }

            // To splay maxInLeft to the root of the left subtree, we can temporarily
            // set the main root to the left subtree's root.
            this.root = leftSubtree;
            splay(maxInLeft); // Now, maxInLeft is the root of the modified left subtree.

            // Join the original right subtree to the new root.
            // After splaying, maxInLeft has no right child.
            this.root.right = rightSubtree;
            if (rightSubtree != null) {
                rightSubtree.parent = this.root;
            }
        }
    }
}